File: src/lexer.cpp
#include <cctype>
#include <iostream>
#include <cassert>
#include <stack>

#include "tokens.hpp"
#include "lexer.hpp"
#include "errors.hpp"

// Prototypes and definitions
// --------------------------
namespace lexer {
    struct OnString {
        char character;
        bool onString;
    };

    struct Source {
        size_t line;
        size_t column;
        size_t index;
        std::filesystem::path path;
        FILE* file_pointer;
        std::stack<OnString> onString;

        Source() {}
        Source(std::filesystem::path path, FILE* file_pointer) : line(1), column(1), index(0), path(path), file_pointer(file_pointer) {}
    };

    char current(Source source) {
        char c = fgetc(source.file_pointer);
        if (c == EOF) fseek(source.file_pointer, 0, SEEK_END);
        else          fseek(source.file_pointer, -1, SEEK_CUR);
        return c;
    }

    void advance(Source& source) {
        if (current(source) == '\n') {
            source.line += 1;
            source.column = 1;
        }
        else {
            source.column++;
        }
        source.index++;
        fgetc(source.file_pointer);
    }

    void advance(Source& source, unsigned int offset) {
        while (offset != 0) {
            advance(source);
            offset--;
        }
    }

    char prev(Source source) {
        int result = fseek(source.file_pointer, -1, SEEK_CUR);
        if (result < 0) return '\n';
        char c = current(source);
        advance(source);
        return c;
    }

    char peek(Source source, unsigned int offset = 1);
    char peek(Source source, unsigned int offset) {
        unsigned int aux = offset;
        while (aux > 0) {
            char c = fgetc(source.file_pointer);
            if (c == EOF) {
                fseek(source.file_pointer, -(int)(offset - aux), SEEK_CUR);
            }
            aux--;
        }
        char c = current(source);
        fseek(source.file_pointer, -(int)offset, SEEK_CUR);
        return c;
    }

    bool at_end(Source source) {
        return current(source) == EOF;
    }

    bool match(Source source, std::string to_match) {
        bool match = true;
        int i = 0;
        while (!at_end(source) && i < to_match.size()) {
            if (current(source) != to_match[i]) {
                match = false;
                break;
            }
            advance(source);
            i++;
        }
        if (at_end(source) && i != to_match.size()) {
            match = false;
        }
        fseek(source.file_pointer, -i, SEEK_CUR);
        return match;
    }

    Result<token::Token, Error> get_token(Source& source);
    Result<token::Token, Error> advance(token::Token token, Source& source, unsigned int offset);
    void advance_until_new_line(Source& source);
    Result<token::Token, Error> get_string(Source& source);
    Result<token::Token, Error> get_number(Source& source);
    Result<token::Token, Error> get_identifier(Source& source);
    std::string get_integer(Source& source);
    Result<token::Token, Error> get_indent(Source& source);
}

// Lexing
// ------
Result<std::vector<token::Token>, Errors> lexer::lex(std::filesystem::path path) {
     std::vector<token::Token> tokens;
     Errors errors;

    // Read file
    FILE* file_pointer = fopen(path.string().c_str(), "rb");
    if (!file_pointer) {
        errors.push_back(errors::file_couldnt_be_found(path));
        return Result<std::vector<token::Token>, Errors>(errors);
    }

    // Create source
    Source source(path, file_pointer);

    // Tokenize
    while (!at_end(source)) {
        auto result = get_token(source);
        if (result.is_ok() && result.get_value() != token::EndOfFile) {
            tokens.push_back(result.get_value());
        }
        else if (result.is_error()) {
            errors.push_back(result.get_error());
            advance(source);
        }
    }
    tokens.push_back(token::Token(token::EndOfFile, source.line, source.column));

    // Close file
    fclose(file_pointer);

    if (errors.size() != 0) return Result<std::vector<token::Token>, Errors>(errors);
    else                    return Result<std::vector<token::Token>, Errors>(tokens);
}

Result<token::Token, Error> lexer::get_token(Source& source) {
    if (at_end(source))      return advance(token::Token(token::EndOfFile, "EndOfFile", source.line, source.column), source, 1);
    if (match(source, "("))  return advance(token::Token(token::LeftParen, "(", source.line, source.column), source, 1);
    if (match(source, ")"))  return advance(token::Token(token::RightParen, ")", source.line, source.column), source, 1);
    if (match(source, "["))  return advance(token::Token(token::LeftBracket, "[", source.line, source.column), source, 1);
    if (match(source, "]"))  return advance(token::Token(token::RightBracket, "]", source.line, source.column), source, 1);
    if (match(source, "{"))  {
        if (!source.onString.empty()) {
            source.onString.push(OnString{'{', false});
        }
        return advance(token::Token(token::LeftCurly, "{", source.line, source.column), source, 1);
    }
    if (match(source, "}")) {
        if (!source.onString.empty()
        && source.onString.top().character == '{') {
            if (source.onString.top().onString) {
                return get_string(source);
            }
            else {
                source.onString.pop();
                return advance(token::Token(token::RightCurly, "}", source.line, source.column), source, 1);
            }
        }

        return advance(token::Token(token::RightCurly, "}", source.line, source.column), source, 1);
    }
    if (match(source, "+"))  return advance(token::Token(token::Plus, "+", source.line, source.column), source, 1);
    if (match(source, "*"))  return advance(token::Token(token::Star, "*", source.line, source.column), source, 1);
    if (match(source, "/"))  return advance(token::Token(token::Slash, "/", source.line, source.column), source, 1);
    if (match(source, "%"))  return advance(token::Token(token::Modulo, "%", source.line, source.column), source, 1);
    if (match(source, ":=")) return advance(token::Token(token::ColonEqual, ":=", source.line, source.column), source, 2);
    if (match(source, ":"))  return advance(token::Token(token::Colon, ":", source.line, source.column), source, 1);
    if (match(source, ","))  return advance(token::Token(token::Comma, ",", source.line, source.column), source, 1);
    if (match(source, "!=")) return advance(token::Token(token::NotEqual, "!=", source.line, source.column), source, 2);
    if (match(source, "==")) return advance(token::Token(token::EqualEqual, "==", source.line, source.column), source, 2);
    if (match(source, "="))  return advance(token::Token(token::Equal, "=", source.line, source.column), source, 1);
    if (match(source, ">=")) return advance(token::Token(token::GreaterEqual, ">=", source.line, source.column), source, 2);
    if (match(source, ">"))  return advance(token::Token(token::Greater, ">", source.line, source.column), source, 1);
    if (match(source, "<=")) return advance(token::Token(token::LessEqual, "<=", source.line, source.column), source, 2);
    if (match(source, "<"))  return advance(token::Token(token::Less, "<", source.line, source.column), source, 1);
    if (match(source, "&"))  return advance(token::Token(token::Ampersand, "&", source.line, source.column), source, 1);
    if (match(source, ".") && isdigit(peek(source))) {
        return get_number(source);
    }
    if (match(source, ".")) {
        return advance(token::Token(token::Dot, ".", source.line, source.column), source, 1);
    }
    if (match(source, "---")) {
        advance(source, 3);
        while (!(at_end(source) || match(source, "---"))) {
            advance(source);
        }
        if (at_end(source)) {
            return Result<token::Token, Error>(std::string("Error: Unclosed block comment\n"));
        }

        advance(source, 3);
        return get_token(source);
    }
    if (match(source, "--")) {
        advance_until_new_line(source);
        return get_token(source);
    }
    if (match(source, "-")) {
        return advance(token::Token(token::Minus, "-", source.line, source.column), source, 1);
    }
    if (match(source, "_")) {
        return get_identifier(source);
    }
    if (match(source, " "))  {
        advance(source);
        return get_token(source);
    }
    if (match(source, "\t")) {
        advance(source);
        return get_token(source);
    }
    if (match(source, "\r\n"))    return advance(token::Token(token::NewLine, "\\n", source.line, source.column), source, 2);
    if (match(source, "\n"))      return advance(token::Token(token::NewLine, "\\n", source.line, source.column), source, 1);
    if (match(source, "\""))      return get_string(source);
    if (isdigit(current(source))) return get_number(source);
    if (isalpha(current(source))) return get_identifier(source);

    return Result<token::Token, Error>(std::string("Error: Unrecognized character \"") + std::string(1, current(source)) + std::string("\".\n"));
}

Result<token::Token, Error> lexer::advance(token::Token token, Source& source, unsigned int offset) {
    advance(source, offset);
    return Result<token::Token, Error>(token);
}

void lexer::advance_until_new_line(Source& source) {
    while (!at_end(source) && current(source) != '\n') {
        advance(source);
    }
    if (!at_end(source)) advance(source);
}

Result<token::Token, Error> lexer::get_string(Source& source) {
    std::string literal = "";
    size_t line = source.line;
    size_t column = source.column;
    bool isRight = false;

    if (current(source) == '\"') {
        source.onString.push(OnString{current(source), true});
    }
    else if (current(source) == '}') {
        source.onString.pop();
        isRight = true;
    }
    else {
        assert(false);
    }

    advance(source);
    while (!(at_end(source) || match(source, "\n"))) {
        if (match(source, "\\n")) {
            literal += "\n";
            advance(source);
        }
        else if (match(source, "\\\"")) {
            literal += "\"";
            advance(source);
        }
        else if (match(source, "\"")) {
            break;
        }
        else if (match(source, "\\{")) {
            literal += "{";
        }
        else if (match(source, "{")) {
            break;
        }
        else {
            literal += current(source);
        }
        advance(source);
    }

    if (at_end(source) || match(source, "\n")) {
        return Result<token::Token, Error>(std::string("Error: Unclosed string\n"));
    }
    else {
        if (match(source, "{")) {
            source.onString.push(OnString{current(source), true});
            advance(source);
            if (isRight) {
                return Result<token::Token, Error>(token::Token(token::StringMiddle, literal, line, column));
            }
            else {
                return Result<token::Token, Error>(token::Token(token::StringLeft, literal, line, column));
            }
        }
        else if (match(source, "\"")) {
            source.onString.pop();
            advance(source);
            if (isRight) {
                return Result<token::Token, Error>(token::Token(token::StringRight, literal, line, column));
            }
            else {
                return Result<token::Token, Error>(token::Token(token::String, literal, line, column));
            }
        }
        else {
            assert(false);
        }
    }
}

Result<token::Token, Error> lexer::get_identifier(Source& source) {
    std::string literal = "";
    size_t line = source.line;
    size_t column = source.column;

    while (!(at_end(source) || match(source, "\n"))) {
        if (!(isalnum(current(source)) || current(source) == '_')) break;
        literal += current(source);
        advance(source);
    }

    if (literal == "if")        return token::Token(token::If, "if", line, column);
    if (literal == "else")      return token::Token(token::Else, "else", line, column);
    if (literal == "while")     return token::Token(token::While, "while", line, column);
    if (literal == "function")  return token::Token(token::Function, "function", line, column);
    if (literal == "interface") return token::Token(token::Interface, "interface", line, column);
    if (literal == "builtin")   return token::Token(token::Builtin, "builtin", line, column);
    if (literal == "type")      return token::Token(token::Type, "type", line, column);
    if (literal == "case")      return token::Token(token::Case, "case", line, column);
    if (literal == "be")        return token::Token(token::Be, "be", line, column);
    if (literal == "true")      return token::Token(token::True, "true", line, column);
    if (literal == "false")     return token::Token(token::False, "false", line, column);
    if (literal == "and")       return token::Token(token::And, "and", line, column);
    if (literal == "or")        return token::Token(token::Or, "or", line, column);
    if (literal == "use")       return token::Token(token::Use, "use", line, column);
    if (literal == "include")   return token::Token(token::Include, "include", line, column);
    if (literal == "break")     return token::Token(token::Break, "break", line, column);
    if (literal == "continue")  return token::Token(token::Continue, "continue", line, column);
    if (literal == "return")    return token::Token(token::Return, "return", line, column);
    if (literal == "mut")       return token::Token(token::Mut, "mut", line, column);
    if (literal == "new")       return token::Token(token::New, "new", line, column);
    if (literal == "not")       return token::Token(token::Not, "not", line, column);
    if (literal == "extern")    return token::Token(token::Extern, "extern", line, column);
    if (literal == "link_with") return token::Token(token::LinkWith, "link_with", line, column);

    return token::Token(token::Identifier, literal, line, column);
}

Result<token::Token, Error> lexer::get_number(Source& source) {
    std::string literal = "";
    size_t line = source.line;
    size_t column = source.column;

    // eg: .8
    if (match(source, ".")) {
        advance(source);
        literal += "." + get_integer(source);
        return token::Token(token::Float, literal, line, column);
    }

    literal += get_integer(source);

    // eg: 6.7
    if (match(source, ".")) {
        advance(source);
        literal += "." + get_integer(source);
        return token::Token(token::Float, literal, line, column);

    // eg: 16
    } else {
        return token::Token(token::Integer, literal, line, column);
    }
}

std::string lexer::get_integer(Source& source) {
    std::string integer = "";
    while (!(at_end(source) || match(source, "\n")) && isdigit(current(source))) {
        integer += current(source);
        advance(source);
    }
    return integer;
}

File: src/lexer.rs
use crate::error::Error;
use crate::token::{Token, TokenKind};

struct Lexer {
    source: Vec<u8>,
    tokens: Vec<Token>,
    errors: Vec<Error>,
    start: usize,
    current: usize,
    line: usize,
    column: usize,
}

pub fn lex(source: String) -> Result<Vec<Token>, Vec<Error>> {
    let mut lexer = Lexer {
        source: source.into_bytes(),
        tokens: Vec::new(),
        errors: Vec::new(),
        start: 0,
        current: 0,
        line: 1,
        column: 1,
    };

    scan_tokens(&mut lexer);
    if lexer.errors.len() == 0 {
        return Ok(lexer.tokens);
    } else {
        return Err(lexer.errors);
    }
}

fn scan_tokens(lexer: &mut Lexer) {
    while !at_end(lexer) {
        scan_token(lexer);
    }

    lexer.tokens.push(Token {
        line: lexer.line,
        column: lexer.column,
        kind: TokenKind::EndOfFile,
    });
}

fn scan_token(lexer: &mut Lexer) {
    lexer.start = lexer.current;
    let char = advance(lexer);
    match char {
        '(' => add_token(lexer, TokenKind::LeftParen),
        ')' => add_token(lexer, TokenKind::RightParen),
        '[' => add_token(lexer, TokenKind::LeftBracket),
        ']' => add_token(lexer, TokenKind::RightBracket),
        '{' => add_token(lexer, TokenKind::LeftCurly),
        '}' => add_token(lexer, TokenKind::RightCurly),
        '+' => add_token(lexer, TokenKind::Plus),
        '*' => add_token(lexer, TokenKind::Star),
        '/' => add_token(lexer, TokenKind::Slash),
        '%' => add_token(lexer, TokenKind::Modulo),
        ',' => add_token(lexer, TokenKind::Comma),
        '&' => add_token(lexer, TokenKind::Ampersand),
        ':' => match matches(lexer, "=") {
            true => add_token(lexer, TokenKind::ColonEqual),
            false => add_token(lexer, TokenKind::Colon),
        },
        '!' => match matches(lexer, "=") {
            true => add_token(lexer, TokenKind::NotEqual),
            _ => {
                lexer.errors.push(Error {
                    message: "Error: Unrecognized character".to_owned(),
                    line: lexer.line,
                    column: lexer.column,
                });
            }
        },
        '=' => match matches(lexer, "=") {
            true => add_token(lexer, TokenKind::EqualEqual),
            false => add_token(lexer, TokenKind::Equal),
        },
        '>' => match matches(lexer, "=") {
            true => add_token(lexer, TokenKind::GreaterEqual),
            false => add_token(lexer, TokenKind::Greater),
        },
        '<' => match matches(lexer, "=") {
            true => add_token(lexer, TokenKind::LessEqual),
            false => add_token(lexer, TokenKind::Less),
        },
        '.' => match peek(lexer).is_digit(10) {
            true => scan_number(lexer),
            false => add_token(lexer, TokenKind::Dot),
        },
        '-' => match matches(lexer, "--") {
            true => {
                while !(at_end(lexer) || matches(lexer, "---")) {
                    advance(lexer);
                }
                if at_end(lexer) {
                    lexer.errors.push(Error {
                        message: "Error: Unclosed block comment".to_owned(),
                        line: lexer.line,
                        column: lexer.column,
                    });
                }
            }
            false => match matches(lexer, "-") {
                true => {
                    advance_until_newline(lexer);
                    advance(lexer);
                }
                false => add_token(lexer, TokenKind::Minus),
            },
        },
        '_' => scan_identifier_or_keyword(lexer),
        ' ' | '\t' => {}
        '\n' => add_token(lexer, TokenKind::NewLine),
        '\r' => {
            if matches(lexer, "\n") {
                add_token(lexer, TokenKind::NewLine)
            } else {
                panic!()
            }
        }
        '\"' => scan_string(lexer),
        _ => {
            if char.is_digit(10) {
                scan_number(lexer);
            } else if char.is_alphanumeric() {
                return scan_identifier_or_keyword(lexer);
            } else {
                lexer.errors.push(Error {
                    message: "Error: Unrecognized character".to_owned(),
                    line: lexer.line,
                    column: lexer.column,
                });
            }
        }
    }
}

fn advance(lexer: &mut Lexer) -> char {
    if peek(lexer) == '\n' || (peek(lexer) == '\r' && peek_next(lexer) == '\n') {
        lexer.line += 1;
        lexer.column = 1;
    } else {
        lexer.column += 1;
    }
    lexer.current += 1;
    return char::from(lexer.source[lexer.current - 1]);
}

fn at_end(lexer: &Lexer) -> bool {
    lexer.current >= lexer.source.len()
}

fn add_token(lexer: &mut Lexer, token: TokenKind) {
    lexer.tokens.push(Token {
        kind: token,
        line: lexer.line,
        column: lexer.column - (lexer.current - lexer.start),
    });
}

fn matches(lexer: &mut Lexer, string: &str) -> bool {
    let mut it_matches = true;
    let mut i = 0;
    while !at_end(lexer) && i < string.len() {
        if peek(lexer) != string.as_bytes()[i] as char {
            it_matches = false;
            lexer.current -= i;
            break;
        }
        advance(lexer);
        i += 1;
    }
    if at_end(lexer) && i != string.len() {
        it_matches = false;
        lexer.current -= i;
    }

    return it_matches;
}

fn peek(lexer: &Lexer) -> char {
    if at_end(lexer) {
        return '\0';
    }
    return lexer.source[lexer.current] as char;
}

fn previous(lexer: &Lexer) -> char {
    if lexer.current > 0 {
        return lexer.source[lexer.current - 1] as char;
    } else {
        return lexer.source[lexer.current] as char;
    }
}

fn peek_next(lexer: &Lexer) -> char {
    if lexer.current + 1 >= lexer.source.len() {
        return '\0';
    }
    return lexer.source[lexer.current + 1] as char;
}

fn scan_string(lexer: &mut Lexer) {
    while peek(lexer) != '"' && !at_end(lexer) {
        if peek(lexer) == '\n' {
            lexer.line += 1
        }
        advance(lexer);
    }

    if at_end(lexer) {
        lexer.errors.push(Error {
            message: String::from("Unterminated string"),
            line: lexer.line,
            column: lexer.column,
        });
        return;
    }

    advance(lexer);

    let value = lexer.source[lexer.start + 1..lexer.current - 1].to_vec();
    add_token(
        lexer,
        TokenKind::String {
            literal: String::from_utf8(value.clone()).unwrap(),
        },
    );
}

fn scan_number(lexer: &mut Lexer) {
    let mut is_float = false;

    // eg: .8
    if previous(lexer) == '.' && peek(lexer).is_digit(10) {
        // consume digits
        while peek(lexer).is_digit(10) {
            advance(lexer);
        }

        is_float = true;
    }
    // eg: 3.5
    else {
        // consume digits
        while peek(lexer).is_digit(10) {
            advance(lexer);
        }

        if matches(lexer, ".") {
            // consume digits
            while peek(lexer).is_digit(10) {
                advance(lexer);
            }

            is_float = true;
        }
    }

    let literal = lexer.source[lexer.start..lexer.current].to_vec();
    let literal = String::from_utf8(literal.clone()).unwrap();
    add_token(
        lexer,
        if is_float {
            TokenKind::Float { literal }
        } else {
            TokenKind::Integer { literal }
        },
    );
}

fn scan_identifier_or_keyword(lexer: &mut Lexer) {
    while peek(lexer).is_alphanumeric() {
        advance(lexer);
    }

    let literal = lexer.source[lexer.start..lexer.current].to_vec();
    let literal = String::from_utf8(literal.clone()).unwrap();

    match &literal[..] {
        "if" => add_token(lexer, TokenKind::If),
        "else" => add_token(lexer, TokenKind::Else),
        "while" => add_token(lexer, TokenKind::While),
        "function" => add_token(lexer, TokenKind::Function),
        "interface" => add_token(lexer, TokenKind::Interface),
        "builtin" => add_token(lexer, TokenKind::Builtin),
        "type" => add_token(lexer, TokenKind::Type),
        "case" => add_token(lexer, TokenKind::Case),
        "be" => add_token(lexer, TokenKind::Be),
        "true" => add_token(lexer, TokenKind::True),
        "false" => add_token(lexer, TokenKind::False),
        "and" => add_token(lexer, TokenKind::And),
        "or" => add_token(lexer, TokenKind::Or),
        "use" => add_token(lexer, TokenKind::Use),
        "include" => add_token(lexer, TokenKind::Include),
        "break" => add_token(lexer, TokenKind::Break),
        "continue" => add_token(lexer, TokenKind::Continue),
        "return" => add_token(lexer, TokenKind::Return),
        "mut" => add_token(lexer, TokenKind::Mut),
        "new" => add_token(lexer, TokenKind::New),
        "not" => add_token(lexer, TokenKind::Not),
        "extern" => add_token(lexer, TokenKind::Extern),
        "link_with" => add_token(lexer, TokenKind::LinkWith),
        _ => add_token(lexer, TokenKind::Identifier { literal }),
    }
}

fn advance_until_newline(lexer: &mut Lexer) {
    while peek(lexer) != '\n' {
        advance(lexer);
    }
}
